skrifa/src/variation.rs

```diff
@@ -0,0 +1,539 @@
+            .unwrap_or(0)
+    }
+
+    /// Returns true if the collection is empty.
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Returns the axis at the given index.
+    pub fn get(&self, index: usize) -> Option<Axis> {
+        let raw = self.fvar.as_ref()?.axes().ok()?.get(index)?.clone();
+        Some(Axis { index, record: raw })
+    }
+
+    /// Returns the axis with the given tag.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use skrifa::prelude::*;
+    /// # fn wrapper(font: &FontRef) {
+    /// let opsz = Tag::new(b"opsz");
+    /// assert_eq!(font.axes().get_by_tag(opsz).unwrap().tag(), opsz);
+    /// # }
+    /// ```
+    pub fn get_by_tag(&self, tag: Tag) -> Option<Axis> {
+        self.iter().find(|axis| axis.tag() == tag)
+    }
+
+    /// Given an iterator of variation settings in user space, computes an
+    /// ordered sequence of normalized coordinates.
+    ///
+    /// * Setting selectors that don't match an axis are ignored.
+    /// * Setting values are clamped to the range of their associated axis
+    ///     before normalization.
+    /// * If more than one setting for an axis is provided, the last one is
+    ///     used.
+    /// * Omitted settings are set to 0.0, representing the default position
+    ///     in variation space.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use skrifa::prelude::*;
+    /// # fn wrapper(font: &FontRef) {
+    /// let location = font.axes().location(&[("wght", 250.0), ("wdth", 75.0)]);
+    /// # }
+    /// ```
+    pub fn location<I>(&self, settings: I) -> Location
+    where
+        I: IntoIterator,
+        I::IntoIter: 'a + Clone,
+        I::Item: Into<VariationSetting>,
+    {
+        let mut location = Location::new(self.len());
+        self.location_to_slice(settings, location.coords_mut());
+        location
+    }
+
+    /// Given an iterator of variation settings in user space, computes an
+    /// ordered sequence of normalized coordinates and stores them in the
+    /// target slice.
+    ///
+    /// * Setting selectors that don't match an axis are ignored.
+    /// * Setting values are clamped to the range of their associated axis
+    ///     before normalization.
+    /// * If more than one setting for an axis is provided, the last one is
+    ///     used.
+    /// * Omitted settings are set to 0.0, representing the default position
+    ///     in variation space.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `location.len() < self.len()`.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use skrifa::prelude::*;
+    /// # fn wrapper(font: &FontRef) {
+    /// let axes = font.axes();
+    /// let mut location = vec![NormalizedCoord::default(); axes.len()];
+    /// axes.location_to_slice(&[("wght", 250.0), ("wdth", 75.0)], &mut location);
+    /// # }
+    /// ```
+    pub fn location_to_slice<I>(&self, settings: I, location: &mut [NormalizedCoord])
```

Comment:
Not sure about Rust conventions, mainly curious: Are there idiomatic function definition styles related to this that inform about a success outcome instead of panicking? Can the reference parameter / output sliced be somehow combined with a `Result<>` type? - but I don't have a strong opinion on that vs. panicking yet.

Reply:
Thanks for the explanation, sgtm to leave as is.

Reply:
I think skrifa should try very very hard to not panic

Reply:
(and therefore so should read-fonts and font-types)